tensorflow_probability/python/mcmc/replica_exchange_mc.py

```diff
@@ -0,0 +1,365 @@
+        next state(s) of the Markov chain(s).
+      kernel_results: A (possibly nested) `tuple`, `namedtuple` or `list` of
+        `Tensor`s representing internal calculations made within this function.
+        This inculdes replica states.
+    """
+    with tf.name_scope(
+        name=mcmc_util.make_name(self.name, 'remc', 'one_step'),
+        values=[current_state, previous_kernel_results]):
+      sampled_replica_states = []
+      sampled_replica_results = []
+      sampled_replica_ratios = []
+      for i in range(self.n_replica):
+        [
+            sampled_state,
+            sampled_results,
+        ] = self.replica_kernels[i].one_step(
+            previous_kernel_results.replica_states[i],
+            previous_kernel_results.replica_results[i])
+        sampled_replica_states.append(sampled_state)
+        sampled_replica_results.append(sampled_results)
+        sampled_replica_ratios.append(self.target_log_prob_fn(sampled_state))
+      sampled_replica_ratios = tf.stack(sampled_replica_ratios, axis=0)
+
+      next_replica_idx = tf.range(self.n_replica)
+      self._seed_stream = distributions_util.gen_new_seed(
+          self._seed_stream, salt='replica_exchange_one_step')
+      exchange_proposed, exchange_proposed_n = \
+          self.exchange_proposed_fn(self.n_replica, seed=self._seed_stream)
+      i = tf.constant(0)
+
+      def cond(i, next_replica_idx):
+        return tf.less(i, exchange_proposed_n)
+
+      def body(i, next_replica_idx):
+        ratio = \
+            sampled_replica_ratios[next_replica_idx[exchange_proposed[i, 0]]] \
+            - sampled_replica_ratios[next_replica_idx[exchange_proposed[i, 1]]]
+        ratio *= self.inverse_temperatures[exchange_proposed[i, 1]] \
+            - self.inverse_temperatures[exchange_proposed[i, 0]]
+        self._seed_stream = distributions_util.gen_new_seed(
+            self._seed_stream, salt='replica_exchange_one_step')
+        log_uniform = tf.log(tf.random_uniform(
+            shape=tf.shape(ratio),
+            dtype=ratio.dtype.base_dtype,
+            seed=self._seed_stream))
+        exchange = log_uniform < ratio
+        exchange_op = tf.sparse_to_dense(
+            [exchange_proposed[i, 0], exchange_proposed[i, 1]],
+            [self.n_replica],
+            [next_replica_idx[exchange_proposed[i, 1]] -
+             next_replica_idx[exchange_proposed[i, 0]],
+             next_replica_idx[exchange_proposed[i, 0]] -
+             next_replica_idx[exchange_proposed[i, 1]]])
+        next_replica_idx = tf.cond(exchange,
+                                   lambda: next_replica_idx + exchange_op,
+                                   lambda: next_replica_idx)
+        return [i + 1, next_replica_idx]
+
+      next_replica_idx = tf.while_loop(
+          cond, body, loop_vars=[i, next_replica_idx])[1]
+
+      next_replica_states = []
+      next_replica_results = []
+      for i in range(self.n_replica):
```

Comment:
Note: since youre using Python control flow (vs TF) we'll need to raise an exception to ensure user meets this contract. (I left comment above.)